torch/distributed/_sharded_tensor/api.py

```diff
@@ -246,6 +262,104 @@ def _init_rpc(self):
+        # from this API.
+        sharded_tensor._sharding_spec = None
+
+        current_rank = dist.get_rank(sharded_tensor._process_group)
+
+        local_shard_metadatas = []
+
+        # collect local shard metadatas from the global sharded_tensor_metadata
+        for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(shard_metadata.placement)
+
+            if current_rank == rank:
+                local_shard_metadatas.append(shard_metadata)
+
+        if len(local_shards) != len(local_shard_metadatas):
+            raise RuntimeError(
+                f'Number of local shards ({len(local_shards)}) does not match number of local '
+                f'shards metadata in sharded_tensor_metadata ({len(local_shard_metadatas)}) '
+                f'on rank ({current_rank}) '
+            )
+
+        for shard in local_shards:
+            shard_meta = shard.metadata
+            local_shard_tensor = shard.tensor
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(shard_meta.placement)
+
+            # validate if shard_meta in the metadatas collected from sharded_tensor_metadata
+            assert shard_meta in local_shard_metadatas, \
+                "local shard metadata not in sharded_tensor_metadata!"
+
+            if shard_meta.shard_lengths != list(local_shard_tensor.size()):
+                raise ValueError(
+                    f'Local shard tensor is incompatible with local ShardMetadata! '
+                    f'local shard tensor size: {local_shard_tensor.size()}, '
+                    f'local ShardMetadata shard lengths: {shard_meta.shard_lengths}'
+                )
+
+            if str(local_shard_tensor.device) != local_device:
+                raise ValueError(
+                    f'Local shard tensor device does not match with local Shard placement! '
+                    f'local shard tensor device: {local_shard_tensor.device}, '
+                    f'local shard metadata placement device: {local_device}'
+                )
+
+            if local_shard_tensor.dtype != sharded_tensor_metadata.dtype:
+                raise ValueError(
+                    f'Local shard tensor dtype does not match with sharded_tensor_metadata! '
+                    f'local shard tensor dtype: {local_shard_tensor.dtype}, '
+                    f'sharded_tensor_metadata dtype: {sharded_tensor_metadata.dtype}'
+                )
+
+            if local_shard_tensor.layout != sharded_tensor_metadata.layout:
+                raise ValueError(
+                    f'Local shard tensor layout does not match with sharded_tensor_metadata! '
+                    f'local shard tensor layout: {local_shard_tensor.dtype}, '
+                    f'sharded_tensor_metadata layout: {sharded_tensor_metadata.layout}'
+                )
+
+            if local_shard_tensor.requires_grad != sharded_tensor_metadata.requires_grad:
```

Comment:
Can't we verify memory layout as well? Since we only support `contiguous`, shouldn't `torch.is_contiguous` be sufficient?